pkg/plugins/policies/meshtrafficpermission/graph/backends/reachable_backend_refs_graph.go

```diff
@@ -0,0 +1,125 @@
+package backends
+
+import (
+	"golang.org/x/exp/maps"
+
+	mesh_proto "github.com/kumahq/kuma/api/mesh/v1alpha1"
+	"github.com/kumahq/kuma/pkg/core"
+	"github.com/kumahq/kuma/pkg/core/resources/apis/mesh"
+	ms_api "github.com/kumahq/kuma/pkg/core/resources/apis/meshservice/api/v1alpha1"
+	core_model "github.com/kumahq/kuma/pkg/core/resources/model"
+	"github.com/kumahq/kuma/pkg/plugins/policies/core/matchers"
+	core_rules "github.com/kumahq/kuma/pkg/plugins/policies/core/rules"
+	mtp_api "github.com/kumahq/kuma/pkg/plugins/policies/meshtrafficpermission/api/v1alpha1"
+	"github.com/kumahq/kuma/pkg/xds/context"
+)
+
+var log = core.Log.WithName("rms-graph")
+
+type BackendKey struct {
+	Kind string
+	Name string
+}
+
+type Graph struct {
+	rules map[BackendKey]core_rules.Rules
+}
+
+func NewGraph() *Graph {
+	return &Graph{
+		rules: map[BackendKey]core_rules.Rules{},
+	}
+}
+
+func (r *Graph) CanReachBackend(fromTags map[string]string, backendRef *mesh_proto.Dataplane_Networking_Outbound_BackendRef) bool {
+	if backendRef.Kind == "MeshExternalService" {
+		return true
+	}
+	rule := r.rules[BackendKey{
+		Kind: backendRef.Kind,
+		Name: backendRef.Name,
+	}].Compute(core_rules.SubsetFromTags(fromTags))
+	if rule == nil {
+		return false
+	}
+	action := rule.Conf.(mtp_api.Conf).Action
+	return action == mtp_api.Allow || action == mtp_api.AllowWithShadowDeny
+}
+
+func BuildRules(meshServices []*ms_api.MeshServiceResource, mtps []*mtp_api.MeshTrafficPermissionResource) map[BackendKey]core_rules.Rules {
```

Comment:
The duplication here is quite significant. This func could basically just be
```
rules := map[BackendKey]core_rules.Rules{}
for _, ms := range meshServices {
		dpTags := maps.Clone(ms.Spec.Selector.DataplaneTags)
		if origin, ok := core_model.ResourceOrigin(ms.GetMeta()); ok {
			dpTags[mesh_proto.ResourceOriginLabel] = string(origin)
		}
		if ms.GetMeta().GetLabels() != nil && ms.GetMeta().GetLabels()[mesh_proto.ZoneTag] != "" {
			dpTags[mesh_proto.ZoneTag] = ms.GetMeta().GetLabels()[mesh_proto.ZoneTag]
		}

rl, ok := ComputeMtpRulesForTags(dpTags, TrimNotSupportedTags(mtps))
		if !ok {
			continue
		}

		rules[BackendKey{
			Kind: "MeshService",
			Name: ms.Meta.GetName(),
		}] = rl

```
you can then call `ComputeMtpRulesForTags` from both graphs.